ThunderCloud/StormLanguageController.swift

```diff
@@ -0,0 +1,410 @@
+            _preferredLocales.insert(Locale(identifier: localeIdentifier), at: 0)
+            overrideLanguage = _overrideLanguage
+        }
+        
+        return _preferredLocales
+    }
+    
+    /// The locales that are available in the language packs
+    private var availableLocales: [LanguagePack]? {
+        
+        let availableLocaleFileNames = ContentController.shared.files(inDirectory: "languages")
+        
+        if let _availableLocaleFileNames = availableLocaleFileNames {
+            
+            return _availableLocaleFileNames.flatMap({ (localeIdentifier: String) -> LanguagePack? in
+                
+                if let languageName = localeIdentifier.components(separatedBy: ".").first {
+                    
+                    let components = languageName.components(separatedBy: "_")
+                    if let _languageString = components.last, let _regionString = components.first {
+                        
+                        if _languageString != _regionString {
+                            let fixedIdentifier = "\(_languageString)_\(_regionString)"
+                            return LanguagePack(locale: Locale(identifier: fixedIdentifier), fileName: languageName)
+                        } else {
+                            return LanguagePack(locale: Locale(identifier: _languageString), fileName: languageName)
+                        }
+                    }
+                }
+                
+                return nil
+            })
+        }
+        return nil
+    }
+    
+    /// Works out the major and regional language packs that are most suitable for the user based on their preferences
+    ///
+    /// - Returns: A tuple containing regional and major language packs. Regional is optional where major should always return one of the packs
+    private func languagePacks() -> (regionalLanguagePack: LanguagePack?, majorLanguagePack: LanguagePack?)? {
+        
+        //Find out if any locales match
+        guard let _availableLocales = availableLocales, let _preferredLocales = preferredLocales else {
+            return nil
+        }
+        
+        var regionalLanguagePack: LanguagePack?
+        var majorLanguagePack: LanguagePack?
+        
+        //Find our language packs that match
+        for pack in _availableLocales {
+            
+            for preferredLocale in _preferredLocales {
+                
+                if preferredLocale.languageCode == pack.locale.languageCode && preferredLocale.regionCode == pack.locale.regionCode {
+                    regionalLanguagePack = pack
+                    
+                    //Set the major language if it matches
+                    if let _languageCode = pack.fileName.components(separatedBy: "_").last {
+                        let languageOnlyLocale = Locale(identifier: _languageCode)
+                        majorLanguagePack = LanguagePack(locale: languageOnlyLocale, fileName: _languageCode)
```

Comment:
The packs that we're looping through are generated from the files in their directory so storing them as file name here is safe because it *does* exist. We later use the ContentController methods to generate a full file path and safely load the file.

I *think* that answers your question